lib/screen/bloc/scan_wallet/scan_wallet_bloc.dart

```diff
@@ -0,0 +1,93 @@
+import 'package:autonomy_flutter/au_bloc.dart';
+import 'package:autonomy_flutter/screen/bloc/scan_wallet/scan_wallet_state.dart';
+import 'package:autonomy_flutter/service/ethereum_service.dart';
+import 'package:autonomy_flutter/service/tezos_service.dart';
+import 'package:autonomy_flutter/util/log.dart';
+import 'package:autonomy_flutter/util/wallet_storage_ext.dart';
+
+class ScanWalletBloc extends AuBloc<ScanWalletEvent, ScanWalletState> {
+  final EthereumService _ethereumService;
+  final TezosService _tezosService;
+
+  ScanWalletBloc(this._ethereumService, this._tezosService)
+      : super(ScanWalletState(addresses: [])) {
+    on<ScanEthereumWalletEvent>(
+      (event, emit) async {
+        emit(state.addNewAddresses([], isScanning: true));
+        List<EthereumAddressInfo> ethereumAddresses = [];
+        int gapCount = 0;
+        int index = event.startIndex;
+        final wallet = event.wallet;
+        while (true) {
+          final address = await wallet.getETHEip55Address(index: index);
+          final balance = await _ethereumService.getBalance(address);
+          final hasBalance = balance.getInWei.compareTo(BigInt.zero) > 0;
+
+          if (hasBalance) {
+            gapCount = 0;
+          } else {
+            gapCount++;
+          }
+          if (hasBalance || event.showEmptyAddresses) {
+            ethereumAddresses.add(EthereumAddressInfo(index, address, balance));
+          }
+
+          final hitStopGap = gapCount >= event.gapLimit;
+          if (hitStopGap || ethereumAddresses.length >= event.maxLength) {
+            if (event.isAdd) {
+              emit(state.addNewAddresses(ethereumAddresses,
+                  hitStopGap: hitStopGap, isScanning: false));
+            } else {
+              emit(ScanWalletState(
+                  addresses: ethereumAddresses, hitStopGap: hitStopGap));
+            }
+            log.info(
+                "ScanEthereumWalletEvent: addresses: $ethereumAddresses hitStopGap: $hitStopGap");
+            break;
+          }
+          index++;
+        }
+      },
+    );
+
+    on<ScanTezosWalletEvent>(
+      (event, emit) async {
+        emit(state.addNewAddresses([], isScanning: true));
+        List<TezosAddressInfo> tezosAddresses = [];
```

Comment:
anyway we can refactor into a function `_scanWallet(CryptoType type)`? See many similar code for both function here.

Reply:
i actually tries that way first, but although they have the same flow, but could not use the same variable, so either 2 similar functions,  or 2 similar blocks of code in one function